# Practicumopdrachten Week 2.3 (eerste kans)

## Dependencies importeren

Kopieer in het codeblok hieronder van het vorige practicum de import-code voor de dependencies die het vaakst worden gebruikt om data in te lezen. Geef er ook de gebruikelijke aliassen aan.<br>
Zet eventuele warnings uit.

In [2]:
import warnings
import pandas as pd
import sqlite3
warnings.simplefilter('ignore')

C:\Users\raoul\AppData\Local\Temp\ipykernel_6996\1599902522.py:2: DeprecationWarning: 
Pyarrow will become a required dependency of pandas in the next major release of pandas (pandas 3.0),
(to allow more performant data types, such as the Arrow string type, and better interoperability with other libraries)
but was not found to be installed on your system.
If this would cause problems for you,
please provide us feedback at https://github.com/pandas-dev/pandas/issues/54466
        
  import pandas as pd


Zet het bestand go_sales.sqlite in hetzelfde mapje als dit notebook.

## Databasetabellen inlezen

Kopieer in het codeblok hieronder van het vorige practicum de code om een connectie met het bestand go_sales.sqlite te maken.

In [9]:
sales_connection = sqlite3.connect('go_sales.sqlite')

Lees van de ingelezen go_sales-database te volgende tabellen in met behulp van "SELECT * FROM *tabel*".
- product
- product_type
- product_line
- sales_staff
- sales_branch
- retailer_site
- country
- order_header
- order_details
- target
- returned_item
- return_reason

In [10]:
product = pd.read_sql_query("SELECT * FROM product", sales_connection)
product_type = pd.read_sql_query("SELECT * FROM product_type", sales_connection)
product_line = pd.read_sql_query("SELECT * FROM product_line", sales_connection)
sales_staff = pd.read_sql_query("SELECT * FROM sales_staff", sales_connection)
sales_branch = pd.read_sql_query("SELECT * FROM sales_branch", sales_connection)
retailer_site = pd.read_sql_query("SELECT * FROM retailer_site", sales_connection)
country = pd.read_sql_query("SELECT * FROM country", sales_connection)
order_header = pd.read_sql_query("SELECT * FROM order_header", sales_connection)
order_details = pd.read_sql_query("SELECT * FROM order_details", sales_connection)
returned_item = pd.read_sql_query("SELECT * FROM returned_item", sales_connection)
return_reason = pd.read_sql_query("SELECT * FROM return_reason", sales_connection)


country



COUNTRY_CODE         COUNTRY LANGUAGE CURRENCY_NAME TRIAL879
0             1          France       EN        francs        T
1            10          Sweden       EN         krona        T
2            11           Japan       EN           yen        T
3            12          Taiwan       EN    new dollar        T
4            13           Korea       EN           won        T
5            14           China       EN      renminbi        T
6            16       Australia       EN       dollars        T
7            17         Belgium       EN        francs        T
8            18         Denmark       EN        kroner        T
9            19           Spain       EN       pesetas        T
10            2         Germany       EN         marks        T
11           20          Mexico       EN         pesos        T
12           21          Brazil       EN         reals        T
13           22         Finland       EN        markka        T
14            3   United States       EN       dollars        T
15            4          Canada       EN       dollars        T
16            5         Austria       EN    schillings        T
17            6           Italy       EN          lira        T
18            7     Netherlands       EN      guilders        T
19            8     Switzerland       EN        francs        T
20            9  United Kingdom       EN        pounds        T

Let op! Voor alle onderstaande opdrachten mag je <u>alleen Python</u> gebruiken, <u>geen SQL!</u>

## Selecties op één tabel zonder functies

Geef een overzicht met daarin de producten en hun productiekosten waarvan de productiekosten lager dan 100 dollar en hoger dan 50 dollar ligt. (2 kolommen, 23 rijen)

In [11]:
kostenSelectie = (product['PRODUCTION_COST'].astype(float) < 100) & (product['PRODUCTION_COST'].astype(float) > 50)

product.loc[(kostenSelectie), ['PRODUCT_NAME', 'PRODUCTION_COST']]

PRODUCT_NAME PRODUCTION_COST
13                Blue Steel Putter              60
14            Blue Steel Max Putter            81.8
17              Course Pro Golf Bag              80
24                  Hibernator Lite              60
25                       Hibernator              86
28  Hibernator Self - Inflating Mat           78.55
31              Hibernator Camp Cot           65.33
32     Canyon Mule Climber Backpack            62.5
56          Granite Climbing Helmet           52.86
58            Husky Harness Extreme              55
69                   Granite Hammer           57.13
71        TrailChef Deluxe Cook Set           85.11
81                        Polar Ice           73.33
83                     Polar Sports              80
84                       Polar Wave              60
85                    Polar Extreme            72.5
88                     Edge Extreme              80
91                        Seeker 35           79.19
92                        Seeker 50           92.58
93           TrailChef Double Flame              75
94                   Seeker Extreme           94.12
97                   Glacier Deluxe           56.25
98                      Glacier GPS           78.55

Geef een overzicht met daarin de producten en hun marge waarvan de marge lager dan 20 % of hoger dan 60 % ligt. (2 kolommen, 7 rijen) 

In [21]:
marginSelectie = (product['MARGIN'].astype(float) < 0.20) | (product['MARGIN'].astype(float) > 0.60)

product.loc[(marginSelectie), ['PRODUCT_NAME', 'MARGIN']]

PRODUCT_NAME MARGIN
29                 Hibernator Pad    .17
32   Canyon Mule Climber Backpack    .17
100             BugShield Natural    .67
101               BugShield Spray    .67
102         BugShield Lotion Lite     .7
103              BugShield Lotion    .63
105             BugShield Extreme    .63

Geef een overzicht met daarin de landen waar met francs wordt betaald. Sorteer de uitkomst op land.  (1 kolom, 3 rijen)

In [27]:
countryFrancs = country['CURRENCY_NAME'] == 'francs'
country.loc[(countryFrancs), ['COUNTRY']]

COUNTRY
0        France
7       Belgium
19  Switzerland

Geef een overzicht met daarin de verschillende introductiedatums waarop producten met meer dan 50% marge worden geïntroduceerd (1 kolom, 7 rijen) 

In [37]:
marginGroterDanVijftig = product['MARGIN'].astype(float) > 0.50
product.loc[(marginGroterDanVijftig), ['INTRODUCTION_DATE', 'MARGIN']].drop_duplicates('INTRODUCTION_DATE')

INTRODUCTION_DATE MARGIN
4          10-12-2019    .51
6          18-12-2019    .51
7          27-12-2019    .55
9           13-1-2020    .55
14         15-12-2019    .55
79         26-10-2016    .51
100         15-2-2011    .67

Geef een overzicht met daarin het eerste adres en de stad van verkoopafdelingen waarvan zowel het tweede adres als de regio bekend is (2 kolommen, 28 rijen)

In [46]:
locatieSelectie = (sales_branch['ADDRESS2'].notna()) & (sales_branch['REGION'].notna())

sales_branch.loc[(locatieSelectie), ['ADDRESS1', 'CITY']]

ADDRESS1              CITY
10  Prol. Paseo de la Reforma No. 51  Distrito Federal
11               202-2-3 Hyakunincho             Tokyo
12                     543-225 Asahi        Osaka City
13                  2315 Queen's Ave         Melbourne
15             Avenida Paulista, 333         São Paulo
21                    3 Albany Court        Birmingham
27               Singelgravenplein 4         Amsterdam

Geef een overzicht met daarin de landen waar met dollars (dollars of new dollar) wordt betaald. Sorteer de uitkomst op land. (1 kolom, 4 rijen) 

In [13]:
dollarOfNew = (country['CURRENCY_NAME'] == 'dollars') | (country['CURRENCY_NAME'] == 'new dollar')
country.loc[dollarOfNew, ['COUNTRY']]


COUNTRY
3          Taiwan
6       Australia
14  United States
15         Canada

Geef een overzicht met daarin beide adressen en de stad van vestigingen van klanten waarvan de postcode begint met een ‘D’ (van duitsland). Filter op vestigingen die een tweede adres hebben. (3 kolommen, 2 rijen) 

In [22]:
DPostcode = sales_branch['POSTAL_ZONE'].str.startswith('D')
sales_branch.loc[DPostcode, ['ADDRESS1', 'ADDRESS2', 'CITY']]

ADDRESS1 ADDRESS2     CITY
0    Schwabentor 35     None  Hamburg
1  Leopoldstraße 36     None  München

## Selecties op één tabel met functies

Geef het totaal aantal producten dat is teruggebracht (1 waarde) 

In [32]:
returnedCount = returned_item['RETURN_CODE'].count()
returnedCount


706

Geef het aantal regio’s waarin verkoopafdelingen gevestigd zijn. (1 waarde)

In [40]:
aantalRegios = sales_branch['REGION'].count()
aantalRegios

15

Maak 3 variabelen:
- Een met de laagste
- Een met de hoogste
- Een met de gemiddelde (afgerond op 2 decimalen)

marge van producten (3 kolommen, 1 rij) 

In [41]:
hoogst = product['MARGIN'].astype(float).max()
laagst = product['MARGIN'].astype(float).min()
gemiddeld = product['MARGIN'].astype(float).mean()
hoogst, laagst, gemiddeld

(0.7, 0.17, 0.396608695652174)

Geef het aantal vestigingen van klanten waarvan het 2e adres niet bekend is (1 waarde)

In [52]:
TweedeNietBekend = sales_branch['ADDRESS2'].isna().sum()
TweedeNietBekend

19

Geef de gemiddelde kostprijs van de verkochte producten waarop korting (unit_sale_price < unit_price) is verleend (1 waarde) 

In [68]:
#Geef de gemiddelde kostprijs van de verkochte producten waarop korting (unit_sale_price < unit_price) is verleend (1 waarde) 


korting = order_details['UNIT_SALE_PRICE'].astype(float) < order_details['UNIT_PRICE'].astype(float)
order_details[korting]['UNIT_PRICE'].astype(float).mean().round(2)


202.23

Geef een overzicht met daarin het aantal medewerkers per medewerkersfunctie (2 kolommen, 7 rijen) 

In [70]:
sales_staff.groupby('POSITION_EN', as_index=False)['SALES_STAFF_CODE'].count()

POSITION_EN  SALES_STAFF_CODE
0                Branch Manager                19
1              District Manager                 4
2               General Manager                 1
3  Level 1 Sales Representative                13
4  Level 2 Sales Representative                39
5  Level 3 Sales Representative                24
6              Regional Manager                 2

Geef een overzicht met daarin per telefoonnummer het aantal medewerkers dat op dat telefoonnummer bereikbaar is. Toon alleen de telefoonnummer waarop meer dan 4 medewerkers bereikbaar zijn. (2 kolommen, 10 rijen) 

In [77]:
filtered_data= sales_staff.groupby('WORK_PHONE', as_index=False)['SALES_STAFF_CODE'].count()
filtered_data = filtered_data[filtered_data['SALES_STAFF_CODE'] > 4]
print(filtered_data)

             WORK_PHONE  SALES_STAFF_CODE
4     +(44) 121 3505267                 5
10  +(61) 03 2982  4242                 5
11  +31 (0)20 692 93 94                 5
13    +33 1 68 94 52 20                 5
18     1 (206) 292-0012                 5
19     1 (305) 557-4810                 6
20     1 (310) 281-5722                 5
21     1 (403) 232-5986                 5
22     1 (416) 493-5595                 5
23     1 (617) 268-6754                 6


## Selecties op meerdere tabellen zonder functies

Geef een overzicht met daarin het eerste adres en de stad van vestigingen van klanten uit ‘Netherlands’ (2 kolommen, 20 rijen) 

In [106]:
joinDf = pd.merge(retailer_site, country, on='COUNTRY_CODE', how='inner')
joinDf[joinDf['COUNTRY'] == 'Netherlands'][['ADDRESS1', 'CITY']]


ADDRESS1          CITY
164                   Beets 36         Beets
165               Perestraat 1      Deventer
166  Laan van Meerdervoort 966      Den Haag
167             Kalfjeslaan 47    Amstelveen
168               Startbaan 11    Amstelveen
169      Noorderspoorsingel 28     Groningen
170              Hogehilweg 13     Amsterdam
171             Spoorstraat 43    Varsseveld
173             Westhavenweg 8     Amsterdam
174              Westzeedijk 4        Muiden
175            Gooiergracht 91         Laren
176            Hoofdstraat 150   Pieterburen
177                Olde Diek 2         Laren
178     Amsterdamsestraatweg 8      Maarssen
179        Krayenhoffstraat 11     Amsterdam
180         Overschiestraat 19     Amsterdam
181              Havenbuurt 12        Marken
182               Duivendaal 1    Wageningen
184              Voorstraat 37  Nieuweschans
185         Stadhouderslaan 18     Rotterdam

Geef een overzicht met daarin de productnamen die tot het producttype ‘Eyewear’ behoren. (1 kolom, 5 rijen) 

In [118]:
productEyewear = pd.merge(product, product_type, on='PRODUCT_TYPE_CODE', how='inner')
productEyewear[productEyewear['PRODUCT_TYPE_EN'] == 'Eyewear'][['PRODUCT_NAME']]

PRODUCT_NAME
80      Polar Sun
81      Polar Ice
83   Polar Sports
84     Polar Wave
85  Polar Extreme

Geef een overzicht met daarin alle unieke eerste adressen van klantvestigingen en de voornaam en achternaam van de verkopers die ‘Branch Manager’ zijn en aan deze vestigingen hebben verkocht (3 kolommen, 1 rij) 

In [130]:
#Geef een overzicht met daarin alle unieke eerste adressen van klantvestigingen en de voornaam en achternaam van de verkopers die ‘Branch Manager’ zijn en aan deze vestigingen hebben verkocht (3 kolommen, 1 rij) 
mergeDf = pd.merge(sales_staff, order_header, on='SALES_STAFF_CODE', how='inner')
mergeDf = pd.merge(mergeDf, retailer_site, on='RETAILER_SITE_CODE', how='inner')
mergeDf = mergeDf[mergeDf['POSITION_EN'] == 'Branch Manager']
mergeDf = mergeDf.drop_duplicates('ADDRESS1')
mergeDf[['ADDRESS1', 'FIRST_NAME', 'LAST_NAME']]


ADDRESS1 FIRST_NAME LAST_NAME
5300  Alameda Santos, 9876     Bayard     Lopes

Geef een overzicht met daarin van de verkopers hun functie en indien zij iets hebben verkocht de datum waarop de verkoop heeft plaatsgevonden. Laat alleen de verschillende namen van de posities zien van de verkopers die het woord ‘Manager’ in hun positienaam hebben staan. (2 kolommen, 9 rijen) 

In [145]:
merge = pd.merge(sales_staff, order_header, on='SALES_STAFF_CODE', how='left')
merge = merge[merge['POSITION_EN'].str.contains('Manager')]
merge.loc[:,['POSITION_EN','ORDER_DATE']].drop_duplicates()


POSITION_EN  ORDER_DATE
421     Branch Manager         NaN
422    General Manager         NaN
1276  Regional Manager         NaN
1643  District Manager         NaN
5325    Branch Manager  2022-09-09
5326    Branch Manager  2022-10-31
5327    Branch Manager  2022-12-13
5328    Branch Manager  2022-03-13
5329    Branch Manager  2022-01-09

Geef een overzicht met daarin de verschillende namen van producten en bijbehorende namen van producttypen van de producten waarvoor ooit meer dan 750 stuks tegelijk verkocht zijn. (2 kolommen, 9 rijen) 

In [152]:
newMerge = pd.merge(product, order_details, on='PRODUCT_NUMBER', how='inner')
newMerge = pd.merge(newMerge, product_type, on='PRODUCT_TYPE_CODE', how='inner')
newMerge = newMerge[newMerge['QUANTITY'] > 750]
newMerge = newMerge[['PRODUCT_NAME', 'PRODUCT_TYPE_EN']].drop_duplicates()
newMerge


PRODUCT_NAME    PRODUCT_TYPE_EN
8616                Star Peg              Tents
15975              Firefly 2           Lanterns
37903        BugShield Spray  Insect Repellents
38313  BugShield Lotion Lite  Insect Repellents
38770       BugShield Lotion  Insect Repellents
39618      BugShield Extreme  Insect Repellents
40754         Sun Shelter 15          Sunscreen
41106         Sun Shelter 30          Sunscreen
41681             Sun Shield          Sunscreen

Geef een overzicht met daarin de productnamen waarvan ooit meer dan 40% korting is verleend. De formule voor korting is: (unit_price - unit_sale_price) / unit_price (1 kolom, 8 rijen) 

In [23]:
kortingMerge = pd.merge(order_details, product, on='PRODUCT_NUMBER', how='inner')
kortingMerge['DISCOUNT'] = ((kortingMerge['UNIT_PRICE'].astype(float) - kortingMerge['UNIT_SALE_PRICE'].astype(float))/kortingMerge['UNIT_PRICE'].astype(float))*100
kortingMerge.loc[kortingMerge['DISCOUNT'] > 40, ['PRODUCT_NAME', 'DISCOUNT']].drop_duplicates('PRODUCT_NAME').round(2)


PRODUCT_NAME  DISCOUNT
5594          BugShield Lotion Lite     51.71
6826              BugShield Extreme     45.43
9413              BugShield Natural     43.33
12276                Sun Shelter 30     41.04
14200              BugShield Lotion     46.12
14985               BugShield Spray     48.00
15124  Hailstorm Titanium Woods Set     40.38
20108                Sun Shelter 15     41.56

Geef een overzicht met daarin de retourreden van producten waarvan ooit meer dan 90% van de aangeschafte hoeveelheid is teruggebracht (return_quantity/quantity). (1 kolom, 3 rijen) 

In [37]:
merged_df = pd.merge(order_details, returned_item, on='ORDER_DETAIL_CODE', how='inner')
merged_df = pd.merge(merged_df, return_reason, on='RETURN_REASON_CODE', how='inner')
merged_df['RETURN%'] = merged_df['RETURN_QUANTITY'].astype(float) / merged_df['QUANTITY'].astype(float) * 100
merged_df.loc[merged_df['RETURN%'] > 90, ['RETURN_DESCRIPTION_EN', 'RETURN%']].drop_duplicates().round(2)

RETURN_DESCRIPTION_EN  RETURN%
0    Wrong product shipped    100.0
1   Unsatisfactory product    100.0
15   Wrong product ordered    100.0

## Selecties op meerdere tabellen met functies

Geef een overzicht met daarin per producttype het aantal producten die tot dat producttype behoren. (2 kolommen, 21 rijen) 

In [46]:
merged_df = pd.merge(product, product_type, on='PRODUCT_TYPE_CODE', how='inner')
merged_df = merged_df.groupby('PRODUCT_TYPE_EN', as_index=False)['PRODUCT_NUMBER'].count()
merged_df

PRODUCT_TYPE_EN  PRODUCT_NUMBER
0             Binoculars               4
1   Climbing Accessories               7
2           Cooking Gear              10
3                Eyewear               5
4              First Aid               5
5       Golf Accessories               4
6      Insect Repellents               5
7                  Irons               4
8                 Knives               5
9               Lanterns              12
10            Navigation               4
11                 Packs               6
12               Putters               3
13                  Rope               4
14                Safety               4
15         Sleeping Bags               7
16             Sunscreen               5
17                 Tents               6
18                 Tools               6
19               Watches               5
20                 Woods               4

Geef een overzicht met daarin per land het aantal vestigingen van klanten die zich in dat land bevinden. (2 kolommen, 21 rijen) 

In [55]:
merged_df = pd.merge(country, retailer_site, on='COUNTRY_CODE', how='inner')
merged_df.groupby('COUNTRY', as_index=False)['RETAILER_SITE_CODE'].count()


COUNTRY  RETAILER_SITE_CODE
0        Australia                  10
1          Austria                  10
2          Belgium                  10
3           Brazil                   1
4           Canada                  31
5            China                  10
6          Denmark                   6
7          Finland                   5
8           France                  29
9          Germany                  32
10           Italy                  12
11           Japan                  25
12           Korea                  10
13          Mexico                  10
14     Netherlands                  20
15           Spain                   9
16          Sweden                  12
17     Switzerland                  13
18          Taiwan                  10
19  United Kingdom                  25
20   United States                 101

Geef een overzicht met daarin van de producten behorend tot het producttype ‘Cooking Gear’ per productnaam de totaal verkochte hoeveelheid en de gemiddelde verkoopprijs. Sorteer de uitkomst op totaal verkochte hoeveelheid. (4 kolommen, 10 rijen) 

In [127]:
merged_df = pd.merge(product, product_type, on='PRODUCT_TYPE_CODE', how='inner')
merged_df = pd.merge(order_details, merged_df, on='PRODUCT_NUMBER', how='inner')

condition = merged_df['PRODUCT_TYPE_EN'] == 'Cooking Gear'

merged_df[condition].groupby('PRODUCT_NAME', as_index=False)['LAST_NAME'].count()

merged_df.loc[condition, ['PRODUCT_NAME', 'QUANTITY', 'UNIT_PRICE', 'UNIT_SALE_PRICE']]




TypeError: agg function failed [how->mean,dtype->object]

Geef een overzicht met daarin van de verkopers behorend tot de verkoopafdeling uit ‘Amsterdam’ de voornaam en achternaam van de verkoper en zijn laagste en hoogste verkooptarget van 2006. (4 kolommen, 4 rijen) 

In [129]:
merged_df = pd.merge(sales_staff, sales_branch, on='SALES_BRANCH_CODE', how='inner')

condition = (merged_df['CITY'] == 'AMSTERDAM')

merged_df[condition].groupby('FIRST_NAME', as_index=False)['LAST_NAME'].count()

Geef een overzicht met daarin per land de naam van het land, de naam van de stad waar de verkoopafdeling is gevestigd (noem de kolomnaam in het overzicht ‘verkoper’) en het aantal steden waar zich klanten bevinden in dat land (noem de kolomnaam in het overzicht ‘klanten’) (3 kolommen, 29 rijen) 

In [131]:
sales_branch

SALES_BRANCH_CODE                          ADDRESS1           ADDRESS2  \
0                 13                    Schwabentor 35               None   
1                 14                  Leopoldstraße 36               None   
2                 15               Isafjordsgatan 30 C               None   
3                 17      7800, 756 - 6th Avenue. S.W.               None   
4                 18                  789 Yonge Street               None   
5                 19            1288 Dorchester Avenue               None   
6                 20                   299 Yale Avenue               None   
7                 21        1288 South Barrington Ave.               None   
8                 22                    10032 NW 186th               None   
9                 23               6c, rue de l'Église               None   
10                24  Prol. Paseo de la Reforma No. 51  Lomas de Santa Fe   
11                25               202-2-3 Hyakunincho        Shinjuku-ku   
12                26                     543-225 Asahi           Asahi-ku   
13                28                  2315 Queen's Ave            Level 2   
14                29     Plaza de la Constitución, s/n               None   
15                30             Avenida Paulista, 333   CJ 231 2o. Andar   
16                31                     Kauppakatu 33               None   
17                32             234-12, Kongdeok-Dong            Mapo-Ku   
18                33      154, Choung-Cheng South Road               None   
19                34          No. 1277 Beijing Xi Road               None   
20                35                    Victoria House    25 Ebury Street   
21                36                    3 Albany Court      Market Street   
22                37                    Birchstraße 13               None   
23                38                 Interleuvenlaan 2               None   
24                39                 Jedleser Straße 7               None   
25                 6     75, rue du Faubourg St-Honoré               None   
26                 7                   Piazza Duomo, 1               None   
27                 9               Singelgravenplein 4      4e verdieping   

                CITY         REGION POSTAL_ZONE COUNTRY_CODE TRIAL888  
0            Hamburg           None     D-22529            2        T  
1            München           None     D-81241            2        T  
2              Kista           None    S-164 40           10        T  
3            Calgary        Alberta     T2P 3Z0            4        T  
4            Toronto        Ontario     M2M 4K8            4        T  
5             Boston  Massachusetts       02133            3        T  
6            Seattle     Washington       98139            3        T  
7        Los Angeles     California       90072            3        T  
8              Miami        Florida       33024            3        T  
9               Lyon           None     F-69002            1        T  
10  Distrito Federal         México       01219           20        T  
11             Tokyo          Tokyo    681-3187           11        T  
12        Osaka City          Osaka    616-5698           11        T  
13         Melbourne            VIC        2088           16        T  
14            Bilbao        Vizcaya       48223           19        T  
15         São Paulo             SP   01403-090           21        T  
16            Kuopio           None   FIN-73100           22        T  
17             Seoul           None     121-020           13        T  
18         San Chung         Taipei         241           12        T  
19          Shanghai           None      200040           14        T  
20            London           None     N17 7LX            9        T  
21        Birmingham  West Midlands     B43 7NP            9        T  
22            Zürich           None     CH-8057            8        T  
23          Heverlee           None      

## Pythonvertalingen van SUBSELECT en UNION met o.a. for-loops

Geef een overzicht met daarin de voornaam en de achternaam van de medewerkers die nog nooit wat hebben verkocht (2 kolommen, 25 rijen) 

In [139]:
merged_df = pd.merge(order_header, sales_staff, on='SALES_STAFF_CODE', how='right')

merged_df.loc[merged_df['ORDER_DATE'].isna(), ['FIRST_NAME', 'LAST_NAME']]

FIRST_NAME      LAST_NAME
421      Walter         Taylor
422    Penelope        Tamrine
607       Polly         Linton
881       Gracy        Gellens
1276       Else         Mörike
1277      Frank          Fuchs
1642      Fritz         Hirsch
1643       Jörg          Kunze
1730      Maria      Iacobucci
1975       Kick        Kalkman
2381      Karin      Bergström
2382      Denis           Pagé
2479      Sally          White
2507      Frank        Bretton
2508    Georges  Saint-Germain
2933       Alex      Rodriguez
3351   Samantha          Floyd
3723      Julie          Olsen
3881      Corey         Wright
4347    Estelle        Johnson
4348       Greg        Belding
4484      Julie        Béchard
4659      Lucía          Reyna
4930     Kazumi        Uragome
5103      Alice         Walter

Geef een overzicht met daarin het aantal producten waarvan de marge lager is dan de gemiddelde marge van alle producten samen. Geef in het overzicht tevens aan wat de gemiddelde marge is van dit aantal producten waarvan de marge lager dan de gemiddelde marge van alle producten samen is. (1 kolom, 2 rijen) 

In [166]:
newDf = product.copy()
counter = 0
product['MARGIN'] = pd.to_numeric(product['MARGIN'], errors='coerce')
gemMargin = product['MARGIN'].mean()

for index, row in product.iterrows():
  marge = product.at[index, 'MARGIN']
  if marge < gemMargin:
    newDf.at[index, 'lagerOfHogerDanGemiddeld'] = 'lager'
    counter += 1
  elif marge > gemMargin:
    newDf.at[index, 'lagerOfHogerDanGemiddeld'] = 'hoger'

newDf['GEM_VAN_LAGER'] = newDf[newDf['lagerOfHogerDanGemiddeld'] == 'lager']['MARGIN'].mean()
newDf['AANTAL_LAGER'] = counter
newDf.loc[:, ['AANTAL_LAGER', 'GEM_VAN_LAGER']].drop_duplicates()

AANTAL_LAGER  GEM_VAN_LAGER
0            59       0.292203

Geef een overzicht met daarin de namen van de producten die voor meer dan 500 (verkoopprijs) zijn verkocht maar nooit zijn teruggebracht. (1 kolom, 13 rijen) 

In [199]:
merged_df = pd.merge(order_details, product, on='PRODUCT_NUMBER', how='inner')
merged_df = pd.merge(merged_df, returned_item, on='ORDER_DETAIL_CODE', how='left')

condition = (merged_df['UNIT_PRICE'].astype(float) > 500) & (merged_df['RETURN_QUANTITY'].isnull())

merged_df.loc[condition, ['PRODUCT_NAME']].drop_duplicates()

PRODUCT_NAME
33               Hailstorm Steel Irons
36            Hailstorm Titanium Irons
41          Lady Hailstorm Steel Irons
44       Lady Hailstorm Titanium Irons
47        Hailstorm Titanium Woods Set
51           Hailstorm Steel Woods Set
55   Lady Hailstorm Titanium Woods Set
58      Lady Hailstorm Steel Woods Set
132                          Star Dome
136                       Star Gazer 2
138                       Star Gazer 3
141                       Star Gazer 6
263                     Husky Rope 200

Geef een overzicht met daarin per (achternaam van) medewerker of hij/zij manager is of niet, door deze informatie toe te voegen als extra 'Ja/Nee'-kolom.<br>
Hint: gebruik een for-loop waarin je o.a. bepaalt of het woord 'Manager' in de functie (position_en) staat. (2 kolommen, 102 rijen).

In [209]:
newSalesStaff = sales_staff.copy()

for index, row in sales_staff.iterrows():
    if 'Manager' in str(row['POSITION_EN']):
        newSalesStaff.at[index, 'IS_MANAGER'] = 'Ja'
    else:
        newSalesStaff.at[index, 'IS_MANAGER'] = 'Nee'

newSalesStaff.loc[:, ['LAST_NAME', 'POSITION_EN','IS_MANAGER']]



LAST_NAME                   POSITION_EN IS_MANAGER
0    Savolainen  Level 2 Sales Representative        Nee
1           Kim  Level 2 Sales Representative        Nee
2          Choi  Level 3 Sales Representative        Nee
3           Kao  Level 2 Sales Representative        Nee
4          Fang  Level 3 Sales Representative        Nee
..          ...                           ...        ...
97     Iglesias  Level 2 Sales Representative        Nee
98        Lopes                Branch Manager         Ja
99       Barros  Level 3 Sales Representative        Nee
100   Guimarães  Level 2 Sales Representative        Nee
101    Lindholm  Level 3 Sales Representative        Nee

[102 rows x 3 columns]

In [210]:
newSalesStaff['IS_MANAGER'] = sales_staff['POSITION_EN'].apply(lambda x: 'Ja' if 'Manager' in x else 'Nee')
newSalesStaff.loc[:, ['LAST_NAME', 'POSITION_EN','IS_MANAGER']]

LAST_NAME                   POSITION_EN IS_MANAGER
0    Savolainen  Level 2 Sales Representative        Nee
1           Kim  Level 2 Sales Representative        Nee
2          Choi  Level 3 Sales Representative        Nee
3           Kao  Level 2 Sales Representative        Nee
4          Fang  Level 3 Sales Representative        Nee
..          ...                           ...        ...
97     Iglesias  Level 2 Sales Representative        Nee
98        Lopes                Branch Manager         Ja
99       Barros  Level 3 Sales Representative        Nee
100   Guimarães  Level 2 Sales Representative        Nee
101    Lindholm  Level 3 Sales Representative        Nee

[102 rows x 3 columns]

Met de onderstaande code laat je Python het huidige jaar uitrekenen.

In [211]:
from datetime import date
date.today().year

2024

Met de onderstaande code selecteer je op een bepaald jaartal uit een datum.

In [216]:
from datetime import datetime

date_str = '16-8-2013'
date_format = '%d-%m-%Y'
date_obj = datetime.strptime(date_str, date_format)

date_obj.year

2013

Geef met behulp van bovenstaande hulpcode een overzicht met daarin op basis van het aantal jaar dat iemand in dienst is of een medewerker ‘kort in dienst’ (minder dan 25 jaar in dienst) of een ‘lang in dienst’ (groter gelijk dan 12 jaar in dienst) is. Geef daarbij per medewerker in een aparte kolom zowel ‘kort in dienst’ als ‘lang in dienst’ aan. Gebruik (wederom) een for-loop.<br>
(2 kolommen, 102 rijen) 

In [224]:
from datetime import datetime
staffCopy = sales_staff.copy()

for index, row in sales_staff.iterrows():
    date_str = staffCopy.at[index, 'DATE_HIRED']
    date_format = '%d-%m-%Y'
    date_obj = datetime.strptime(date_str, date_format)

    if (date.today().year - date_obj.year) < 25:
        staffCopy.at[index, 'IN_DIENST'] = 'Kort in dienst'
    elif (date.today().year - date_obj.year) >= 25:
        staffCopy.at[index, 'IN_DIENST'] = 'Lang in dienst'

staffCopy.loc[:, ['LAST_NAME', 'DATE_HIRED', 'IN_DIENST']]


LAST_NAME  DATE_HIRED       IN_DIENST
0    Savolainen   23-7-2014  Kort in dienst
1           Kim  15-12-2014  Kort in dienst
2          Choi  27-11-2014  Kort in dienst
3           Kao   7-10-2014  Kort in dienst
4          Fang    8-5-2012  Kort in dienst
..          ...         ...             ...
97     Iglesias   19-2-2014  Kort in dienst
98        Lopes   12-8-2013  Kort in dienst
99       Barros   10-2-2017  Kort in dienst
100   Guimarães   14-2-2015  Kort in dienst
101    Lindholm   17-8-2013  Kort in dienst

[102 rows x 3 columns]